[Chng Wei Cheng] iP

AnAfterthought opened the pull request:
## BPlusChatter 🤖 
- Easy to use
- *Fast*
- Text-based
- **FREE**

> [!IMPORTANT]
> How to install

Simply:
1. Download from [here](https://github.com/AnAfterthought/ip).
2. Run it using Java.
3. Enter your tasks.

Features:
- [x] Task creation
- [x] Saving tasks
- [ ] Reminders

If you wish to change save files, simply change it in the `Main` class:
```java
public class Main extends Application {

    private BPlusChatter bPlusChatter = new BPlusChatter("./data/BPlusChatter.txt"); //change the file path here

    @Override
    public void start(Stage stage) {
        try {
            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
            AnchorPane ap = fxmlLoader.load();
            Scene scene = new Scene(ap);
            stage.setScene(scene);
            stage.setTitle("BPlusChatter");
            fxmlLoader.<MainWindow>getController().setBPlusChatter(bPlusChatter);  // inject the BPlusChatter instance
            stage.show();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

AnAfterthought committed:
Rename main class

AnAfterthought committed:
Add greeting and exit

AnAfterthought committed:
Added support for echo

AnAfterthought committed:
Added add and list features

AnAfterthought committed:
Added Task class

AnAfterthought committed:
Add task feature now uses Task objects

AnAfterthought committed:
Add support for mark and unmark

AnAfterthought committed:
Add ToDo class

AnAfterthought committed:
Add Deadline class

AnAfterthought committed:
Add Event class

AnAfterthought committed:
Performed I/O redirection testing

AnAfterthought committed:
Add exception handling for unknown commands

AnAfterthought committed:
Add exception handing for ToDo command

AnAfterthought committed:
Add exception handling for deadline command

AnAfterthought committed:
Add exception handling for event command

AnAfterthought committed:
Add exception handling for mark command

AnAfterthought committed:
Add exception handling for unmark command

AnAfterthought committed:
Add bye command

AnAfterthought committed:
Add I/O testing for exception handling

AnAfterthought committed:
Update list and add to use ArrayList

AnAfterthought committed:
Update mark and unmark to use ArrayList

AnAfterthought committed:
Add support for delete

AnAfterthought committed:
Add exception handling for delete

AnAfterthought committed:
Update exceptions to print error messages instead of main program

AnAfterthought committed:
Add I/O testing for delete

AnAfterthought committed:
Remove unneeded imports

AnAfterthought committed:
Remove separators

AnAfterthought committed:
Update to use switch statements

AnAfterthought committed:
Add support for save and load feature

Chatbot does not save tasks in the hard disk. It does not retrieve
tasks from hard disk.

Chatbot should be able to save and load tasks so that users can
access them when the chatbot is restarted.

Write tasks into a text file after each update to the tasklist. When
the chatbot starts up, retrieve tasks from file.

Saving tasks into a text file and retrieving tasks from a text file
is simple. The tasks are readable directly from the file.

AnAfterthought committed:
Merge branch 'branch-Level-7'

AnAfterthought committed:
Add support for reading dates and times

Chatbot is not able to understand strings in date and time format.

It is easier to understand date and time format

Add functionality to read date and time in the yyyy-MM-dd HHmm format.
Add functionality to print date and time in the MMM d yyyy hh:mm am/pm
format

These formats are easy to use and read

AnAfterthought committed:
Add support for on command to find tasks on specific date

AnAfterthought committed:
Update I/O tests to use new date and time format

AnAfterthought committed:
Merge branch 'branch-Level-8'

AnAfterthought committed:
Add classes Ui, Storage, Parser, Tasklist

Many of the chatbot's features are performed in the main class.

Using OOP will make the code neater and easier to test

Let's
* Create Ui class to handle user interactions
* Create Storage class to handle saving and loading of tasks
* Create Parser class to handle commands
* Create TaskList to handle task list modifications

Grouping the features into different classes makes it easier to test
and debug. It also allows for polymorphism if needed.

AnAfterthought committed:
Add load feature to Storage class

AnAfterthought committed:
Add support to adding tasks in Parser class

AnAfterthought committed:
Add support for mark, unmark and delete command in Parser class

AnAfterthought committed:
Add support for saving feature to Storage class

AnAfterthought committed:
Merge branch 'branch-A-MoreOOP'

AnAfterthought committed:
Organize classes into packages

All the classes are in src/main/java

This makes it difficult to manage classes

Let's
* Move exceptions to exception package
* Move Ui class and chatbot class to ui package
* Move Parser class to command package
* Move task classes to task package
* Move Storage class to storage package

This makes it easier to manage different classes

AnAfterthought committed:
Merge branch 'branch-A-Packages'

AnAfterthought committed:
Merge branch 'add-gradle-support'

AnAfterthought committed:
Set up gradle to run chatbot

AnAfterthought committed:
Add JUnit tests for TaskList and Parser

The chatbot has no JUnit tests

Adding JUnit tests will make debugging the code easier. It will
automate the testing.

Let's
* Add JUnit tests to test TaskList features add, delete and get
* Add JUnit tests to test Parser commands (e.g. parseToDo())

JUnit tests allow for automated testing

AnAfterthought committed:
Add JavaDoc comments

AnAfterthought committed:
Update showAdd function in Ui class to take in an integer instead of a TaskList object

AnAfterthought committed:
Add support for find command

AnAfterthought committed:
Merge branch 'branch-A-JavaDoc'

AnAfterthought committed:
Merge branch 'branch-A-CodingStandard'

AnAfterthought committed:
Merge branch 'branch-Level-9'

AnAfterthought committed:
Add support to get isDone value of task

AnAfterthought committed:
Update code to follow sample checkstyle

AnAfterthought committed:
Merge branch 'branch-A-Checkstyle'

AnAfterthought committed:
Add GUI using JavaFX

Chatbot is CLI based.

CLI based chatbot is less visually appealing that GUI. Having a GUI
makes it more accessible.

Implement GUI using JavaFX. The commands remain command line based.

JavaFX is used to create GUIs in Java. The commands are command
line based as it is faster to type everything in a single line than
filling in multiple blanks.

AnAfterthought committed:
Merge branch 'branch-Level-10'

AnAfterthought committed:
Update parseFind to use varargs

AnAfterthought committed:
Merge branch 'branch-A-Varargs'

davidchongg commented on src/main/java/bpluschatter/task/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+     * @return String to be saved in file.
+     */
+    @Override
+    public String toFileFormat() {
```
Good job of using verbs for naming methods.

davidchongg commented on src/main/java/bpluschatter/task/TaskList.java:
```diff
@@ -0,0 +1,85 @@
+     *
+     * @return Number of tasks.
+     */
+    public int size() {
```
Perhaps could rename as getSize()?

davidchongg commented on src/main/java/bpluschatter/task/TaskList.java:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * Prints all tasks.
+     */
+    public void list() {
```
Perhaps could rename as printList() instead since it prints instead of returning list only.

Emmanuel2001 commented on src/main/java/bpluschatter/task/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+/**
+ * Represents a deadline task
+ */
```
I like how you have header comments for each public class! Nice!

Emmanuel2001 commented on src/main/java/bpluschatter/command/Parser.java:
```diff
@@ -0,0 +1,290 @@
+        if (taskParts.length == 2) {
+            details = taskParts[1];
+        }
+        try {
+            switch (command.toLowerCase()) {
+            case "list" -> ui.setList(tasks);
+            case "todo" -> {
+                return parseToDo(details, tasks, ui);
+            }
+            case "deadline" -> {
+                return parseDeadline(details, tasks, ui);
+            }
+            case "event" -> {
+                return parseEvent(details, tasks, ui);
+            }
+            case "mark" -> {
+                return parseMark(details, tasks, ui, true);
+            }
+            case "unmark" -> {
+                return parseMark(details, tasks, ui, false);
+            }
+            case "delete" -> {
+                return parseDelete(details, tasks, ui);
+            }
+            case "on" -> parseOn(details, tasks);
+            case "find" -> parseFind(tasks, ui, details.split(","));
```
The code block for the switch cases looks neat, with each function catering to each features, nice work!

Emmanuel2001 commented on src/main/java/bpluschatter/ui/Ui.java:
```diff
@@ -0,0 +1,187 @@
+    private String message;
+    private boolean isError = false;
+
+    /**
+     * Sets welcome message.
+     */
+    public void setWelcome() {
+        this.message = "Hello! I'm BPlusChatter :)\nWhat can I do for you?";
+        isError = false;
+    }
```
Method names follow the naming convention and describes what it does. Well Done!

Emmanuel2001 commented on src/main/java/bpluschatter/task/Task.java:
```diff
@@ -0,0 +1,76 @@
+ */
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
I like how the boolean is named like booleans!

Emmanuel2001 commented on src/main/java/bpluschatter/command/Parser.java:
```diff
@@ -0,0 +1,290 @@
+            throw new InvalidEventException();
+        }
+
+        String[] detailParts = details.split(" /from ", 2);
+        if (detailParts.length != 2) {
+            throw new InvalidEventException();
+        }
+        String[] duration = detailParts[1].split(" /to ", 2);
+        if (duration.length != 2) {
+            throw new InvalidEventException();
+        }
```
I like how you handle the errors straight, nice!

Emmanuel2001 reviewed (commented):
Hi Wei Cheng, you did a good job! I can tell that you have put in a lot of effort into ensuring that your code is of good quality and standard :)

AnAfterthought committed:
Add assertions

The code does not have any assertions.

There is no way to check if functions that do not return null
encounter a situation where they return null. Additionally,
there is no way to ensure the code does not go to areas it
should not go to.

Let's
* Assert that objects returned from functions are not null
* Assert that code does not reach outside of switch statements if
it does not have to

Using assertions will allow for early detection of bugs.

AnAfterthought committed:
Improve code quality

AnAfterthought committed:
Implement streams for finding tasks

Tasks matching conditions, such as those containing specific
keywords, are found using nested for loops and conditionals.

Nested for loops result in arrowhead style code which reduces
readability.

Let's
* implement a method to convert TaskList to a stream
* implement boolean methods to test for conditions
* use streams to filter tasks matching conditions

Using streams makes the code more readable and easier to
understand.

AnAfterthought committed:
Merge pull request #1 from AnAfterthought/branch-A-Assertions

Add assertions

AnAfterthought committed:
Merge branch 'master' into branch-A-Streams

AnAfterthought committed:
Merge pull request #3 from AnAfterthought/branch-A-Streams

Implement streams for finding tasks

AnAfterthought committed:
Fix bug where list would assert false

AnAfterthought committed:
Merge branch 'master' into branch-A-CodeQuality

AnAfterthought committed:
Merge pull request #2 from AnAfterthought/branch-A-CodeQuality

Improve code quality

AnAfterthought committed:
Add support for Continuous Integration

AnAfterthought committed:
Add support for task priority

Tasks do not have priority levels.

Being able to prioritize tasks to different levels allows users to
decide which tasks are more important.

Let's,
* add enumerations for priority levels HIGH, MEDIUM and LOW
* update commands to allow addition of priority levels
* sort tasks by priority level, starting from HIGH priority

Users can easily set priority to their tasks. When they list their
tasks, they can easily see which ones are more important.

AnAfterthought committed:
Merge pull request #4 from AnAfterthought/branch-C-Priority

Add support for task priority

AnAfterthought committed:
Add user guide

AnAfterthought committed:
Merge pull request #6 from AnAfterthought/branch-A-UserGuide

Add user guide

AnAfterthought committed:
Update user guide

AnAfterthought committed:
Update user guide with more information picture

AnAfterthought committed:
Remove I/O redirection tests from CI

AnAfterthought committed:
Update descriptions to show lowercase am/pm for time

AnAfterthought committed:
Update descriptions to show lowercase am/pm for time

AnAfterthought committed:
Update descriptions to show lowercase am/pm for event start time

AnAfterthought committed:
Add support for handling corrupted save file

AnAfterthought committed:
Update javadocs

AnAfterthought committed:
Update user guide

AnAfterthought committed:
Update javadocs

AnAfterthought committed:
Update user guide to add save information

AnAfterthought committed:
Update user guide formatting

AnAfterthought committed:
Rename variables

AnAfterthought committed:
Add JUnit tests for Deadline and Event

AnAfterthought committed:
Update user guide

AnAfterthought committed:
Update Ui.png to be cleaner
